ext_mongodb.php

```diff
@@ -378,65 +378,229 @@ public function __debugInfo()
-					throw new \MongoDB\Driver\Exception\InvalidArgumentException(
-						'Expected "readConcern" option to be MongoDB\Driver\ReadConcern, ' .
-						gettype($options['readConcern']) . ' given'
-					);
-				} else {
-					if ($options['readConcern']->getLevel() != NULL) {
-						$this->query['readConcern'] = $options['readConcern']->getLevel();
-					}
-				}
+		return false;
+	}
+
+	private function _queryOptInt64( string $optName, array $options, string $key )
+	{
+		if ( array_key_exists( $key, $options ) )
+		{
+			$this->opts[$optName] = (int) $options[$key];
+			return true;
+		}
+
+		return false;
+	}
+
+	private function _queryOptString( string $optName, array $options, string $key )
+	{
+		if ( array_key_exists( $key, $options ) )
+		{
+			if ( !is_string( $options[$key] ) )
+			{
+				throw new \MongoDB\Driver\Exception\InvalidArgumentException(
+					sprintf(
+						'Expected "%s" %s to be string, %s given',
+						$key, $key[0] == '$' ? 'modifier' : 'option', gettype( $options[$key] )
+					)
+				);
 			}
+			$this->opts[$optName] = $options[$key];
+			return true;
+		}
 
-			if (array_key_exists('modifiers', $options)) {
-				Utils::mustBeArrayOrObject('modifiers', $options['modifiers']);
-				foreach ($options['modifiers'] as $key => $value) {
-					$this->query['query'][$key] = $value;
-				}
+		return false;
+	}
+
+	/* Helpers for setting / validating complex options */
+	private function _queryInitHint( array $options, array $modifiers )
+	{
+		if ( array_key_exists( 'hint', $options ) )
+		{
+			if ( is_string( $options['hint'] ) )
+			{
+				$this->_queryOptString( 'hint', $options, 'hint' );
+			}
+			else if ( is_object( $options['hint'] ) || is_array( $options['hint'] ) )
+			{
+				$this->_queryOptDocument( 'hint', $options, 'hint' );
+			}
+			else
+			{
+				throw new \MongoDB\Driver\Exception\InvalidArgumentException(
+					sprintf(
+						'Expected "hint" option to be string, array, or object, %s given',
+						gettype( $options['hint'] )
+					)
+				);
+			}
+		}
+		else if ( array_key_exists( '$hint', $modifiers ) )
+		{
+			if ( is_string( $modifiers['$hint'] ) )
+			{
+				$this->_queryOptString( 'hint', $modifiers, '$hint' );
 			}
+			else if ( is_object( $modifiers['$hint'] ) || is_array( $modifiers['$hint'] ) )
+			{
+				$this->_queryOptDocument( 'hint', $modifiers, '$hint' );
+			}
+			else
+			{
+				throw new \MongoDB\Driver\Exception\InvalidArgumentException(
+					sprintf(
+						'Expected "$hint" modifier to be string, array, or object, %s given',
+						gettype( $modifiers['$hint'] )
+					)
+				);
+			}
+		}
+	}
 
-			if (array_key_exists('projection', $options)) {
-				Utils::mustBeArrayOrObject('projection', $options['projection']);
-				$this->query['fields'] = (array) $options['projection'];
+	private function _queryInitLimitAndSingleBatch( array $options )
+	{
+		if ( array_key_exists( 'limit', $options ) && $options['limit'] < 0 )
+		{
+			$this->opts['limit'] = 0 - $options['limit'];
```

Comment:
This should still perform an integer cast, as `0 - (-1.5)` will result in `1.5`.